validator/sawtooth_validator/journal/completer.py

```diff
@@ -36,93 +39,266 @@ class Completer(object):
         """
         self.gossip = gossip
-        self.batch_cache = TimedCache()
-        self.block_cache = BlockCache(block_store)
+        self.batch_cache = TimedCache(cache_purge_frequency)
+        self.block_cache = BlockCache(block_store, cache_purge_frequency)
         # avoid throwing away the genesis block
         self.block_cache[NULL_BLOCK_IDENTIFIER] = None
+        self._seen_txns = TimedCache(cache_purge_frequency)
+        self._incomplete_batches = TimedCache(cache_purge_frequency)
+        self._incomplete_blocks = TimedCache(cache_purge_frequency)
         self._on_block_received = None
         self._on_batch_received = None
+        self.lock = RLock()
+        self._cache_purge_frequency = cache_purge_frequency
+        self._purge_time = time.time() + self._cache_purge_frequency
 
     def _complete_block(self, block):
         """ Check the block to see if it is complete and if it can be passed to
             the journal. If the block's predecessor is not in the block_cache
-            the predecessor is requested and the current block is dropped.
-            False is returned. If the block.batches and block.header.batch_ids
-            are not the same length, False is returned. The block's batch list
-            needs to be in the same order as the block.header.batch_ids list.
-            If any batches are missing from the block and we do not have the
-            batches in the batch_cache, they are requested. The block is then
-            dropped and False is returned. If the block has all of its expected
-            batches it is added to the block_cache and True is returned.
+            the predecessor is requested and the current block is added to the
+            the incomplete_block cache. If the block.batches and
+            block.header.batch_ids are not the same length, the batch_id list
+            is checked against the batch_cache to see if the batch_list can be
+            built. If any batches are missing from the block and we do not have
+            the batches in the batch_cache, they are requested. The block is
+            then added to the incomplete_block cache. If we can complete the
+            block, a new batch list is created in the correct order and added
+            to the block. The block is now considered complete and is returned.
+            If block.batches and block.header.batch_ids are the same length,
+            the block's batch list needs to be in the same order as the
+            block.header.batch_ids list. If the block has all of its expected
+            batches but are not in the correct order, the batch list is rebuilt
+            and added to the block. Once a block has the correct batch list it
+            is added to the block_cache and is returned.
+
         """
-        valid = True
+
+        if block.header_signature in self.block_cache:
+            LOGGER.debug("Drop duplicate block: %s", block)
+            return
```

Comment:
These should return None explicitly in this function, since the return values are the complete block or None